theano/gof/tests/test_wrapper.py

```diff
@@ -0,0 +1,245 @@
+            NpyIter* iterator = NpyIter_New(tensor,
+                NPY_ITER_READWRITE | NPY_ITER_EXTERNAL_LOOP | NPY_ITER_REFS_OK,
+                NPY_KEEPORDER, NPY_NO_CASTING, NULL);
+            if(iterator == NULL) {
+                PyErr_SetString(PyExc_RuntimeError, "Unable to iterate over a tensor for an elemwise operation.");
+                return -1;
+            }
+            NpyIter_IterNextFunc* get_next = NpyIter_GetIterNext(iterator, NULL);
+            char** data_ptr = NpyIter_GetDataPtrArray(iterator);
+            npy_intp* stride_ptr = NpyIter_GetInnerStrideArray(iterator);
+            npy_intp* innersize_ptr = NpyIter_GetInnerLoopSizePtr(iterator);
+            do {
+                char* data = *data_ptr;
+                npy_intp stride = *stride_ptr;
+                npy_intp count = *innersize_ptr;
+                while(count) {
+                    %(float_type)s x = *((%(float_type)s*)data);
+                    *((%(float_type)s*)data) = a*x*x + b*x + c;
+                    data += stride;
+                    --count;
+                }
+            } while(get_next(iterator));
+            NpyIter_Deallocate(iterator);
+            return 0;
+        }
+        """ % {'float_type': float_type}
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        X = inputs[0]
+        Y = outputs[0]
+        coeff = sub['params']
+        fail = sub['fail']
+        float_type = node.inputs[0].type.dtype_specs()[1]
+        float_typenum = numpy.dtype(node.inputs[0].type.dtype).num
+        coeff_type = 'npy_' + numpy.dtype(dtype).name
+        return """
+        %(float_type)s a = (%(float_type)s) (*(%(coeff_type)s*) PyArray_GETPTR1(%(coeff)s->a, 0)); // 0-D TensorType.
+        %(float_type)s b =                                                      %(coeff)s->b;      // Scalar.
+        %(float_type)s c =                     (%(float_type)s)PyFloat_AsDouble(%(coeff)s->c);     // Generic.
+        Py_XDECREF(%(Y)s);
+        %(Y)s = (PyArrayObject*)PyArray_EMPTY(PyArray_NDIM(%(X)s), PyArray_DIMS(%(X)s), %(float_typenum)s, PyArray_IS_F_CONTIGUOUS(%(X)s));
+        if (PyArray_CopyInto(%(Y)s, %(X)s) != 0) {
+            PyErr_SetString(PyExc_RuntimeError, "Unable to copy input into output.");
+            %(fail)s
+        };
+        if (quadratic_%(float_type)s(%(Y)s, a, b, c) != 0) {
+            PyErr_SetString(PyExc_RuntimeError, "Unable to compute quadratic function.");
+            %(fail)s
+        }
+        """ % locals()
+
+
+# Same op as above, but implemented as a COp (with C code in an external file).
+class QuadraticCOpFunc(COp):
+    __props__ = ('a', 'b', 'c')
+    params_type = Wrapper(a=tensor_type_0d,
+                          b=scalar_type,
+                          c=generic_type)
+
+    def get_op_params(self):
+        return [('QUADRATIC_WRAPPER', self.params_type.name),
```

Comment:
About macros, can I suggest to extend the abstract class `Type` (precisely `CLinkerType`) ? I think about a new method:
 * `c_dtype()` which should return the primitive C type of items into a variable handled by the type. For example, with `TensorType(dtype='int64' ...)`, it should return `npy_int64`. For `GpuArrayType(dtype='int32', ...)`, it should return `ga_long`.

Which such a contract, if `c_dtype()` is implemented for tthe type of a param named `key` into Wrapper, I could automatically add a default local macro like `DTYPE_PARAM_key`, which could help when we have static conversions to do into C files for COps. If `c_dtype()` is not implemented (e.g. for `Generic()`), then the macro is not defined (developper will have to deal with types dynamically).

@nouiz @abergeron